### PR TITLE
#5072: Marketplace Activation Followup: bug fix and types cleanup

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Part of #5072, fixing a bug with the activate recipe sidebar tab when there are other tabs in the sidebar
- Cleans up some stuff with types across the whole sidebar app

## Reviewer Tips

Main changes are:
- `type SidebarEntry = PanelEntry | TemporaryPanelEntry | FormEntry | ActivateRecipeEntry`
- `PanelEntry`, `TemporaryPanelEntry`, `FormEntry`, and `ActivateRecipeEntry` have a `type` field now to differentiate them
  - Test factories set up to generate SidebarEntry instances, tests refactored
- `mapTabEventKey()` renamed to `eventKeyForEntry()`, updated to handle activate-recipe tabs as well
- In `sidebarSlice`, `eventKeyExists()` updated to include activate-recipe keys to fix activate tab bug in the UI
- `displayTemporaryInfo()`
  - Typing fixed to use `TemporaryPanelEntry` instead of `PanelEntry` throughout
  - `entry/refreshEntry()` props condensed into `getPanelEntry()` since the usage is consistent at both call-sites, a factory function is created and then called to generate the entry prop in both places, so move this inside the display function.

## Checklist

- [x] Add tests
- [ ] Designate a primary reviewer

### CODE REVIEW

File: src/testUtils/factories.ts [hunk 1]
```diff
@@ -768,3 +777,87 @@ export const marketplaceListingFactory = define<MarketplaceListing>({
...
+): TemporaryPanelEntry;
+export function sidebarEntryFactory(
+  type: "form",
+  override?: FactoryConfig<FormEntry>
+): FormEntry;
+export function sidebarEntryFactory(
+  type: "activateRecipe",
+  override?: FactoryConfig<ActivateRecipeEntry>
+): ActivateRecipeEntry;
+export function sidebarEntryFactory(
```
Comment (twschiller) [inline src/testUtils/factories.ts:837]: This seems unnecessary vs. just having the factory for each. But I'm OK keeping

File: src/sidebar/sidebarSlice.ts [hunk 2]
```diff
@@ -151,6 +156,16 @@ async function resolvePanel(
   resolveTemporaryPanel(topLevelFrame, nonce, action);
 }
 
+function fixActiveTabOnRemove(
+  state: WritableDraft<SidebarState>,
+  removedEntry: SidebarEntry
+) {
+  // Only update the active panel if the panel needs to change
+  if (state.activeKey === eventKeyForEntry(removedEntry)) {
```
Comment (twschiller) [inline src/sidebar/sidebarSlice.ts:164]: To be defensive, I think this should handle removedEntry as being null. I'm not confident there's no corner cases where in the reducer an entry gets double-removed

File: src/sidebar/sidebarSlice.test.ts [hunk 3]
```diff
@@ -63,31 +61,28 @@ describe("sidebarSlice.selectTab", () => {
...
 });
 
 describe("sidebarSlice.addTemporaryPanel", () => {
-  it("cancels existing temporary panel", async () => {
-    const existingPanel = { nonce: uuidv4(), extensionId: uuidv4() };
-    const otherExistingPanel = { nonce: uuidv4(), extensionId: uuidv4() };
-    const newPanel: TemporaryPanelEntry = {
-      nonce: uuidv4(),
+  it("cancels existing temporary panel for extension", async () => {
+    const existingPanel = sidebarEntryFactory("temporaryPanel");
```
Comment (twschiller) [inline src/sidebar/sidebarSlice.test.ts:81]: 👍

**Review by twschiller — APPROVED** [review]

File: src/testUtils/factories.ts (hunk 1 above)
Comment (BLoe) [reply src/testUtils/factories.ts:837]: I think at one point I thought I was going to be able to get away with not passing the type string into the factory at all, but you're right, this isn't totally necessary as it is now.

Comment (BLoe) [reply src/testUtils/factories.ts:837]: I think at one point I thought I was going to be able to get away with not passing the type string into the factory at all, but you're right, this isn't totally necessary as it is now.
